File: tiq-e/linked-list/2/main.ts
import { performance } from "perf_hooks";

class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  if (head == null) return null;

  const h0 = new ListNode(0, head);
  let n1: ListNode | null = h0;
  let n2: ListNode | null = h0;

  for (let i = 1; i <= n + 1; i++) {
    n1 = n1.next!;
  }

  while (n1 != null) {
    n1 = n1.next!;
    n2 = n2.next!;
  }

  n2.next = n2.next!.next;
  return h0.next;
}

const head = new ListNode(1, null);

const start = performance.now();
const result = removeNthFromEnd(head, 1);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);

File: tiq-e/linked-list/2/main.rs
use std::time::Instant;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode {
            val: -1,
            next: head,
        });

        let mut right = dummy.clone();
        let mut left = dummy.as_mut();

        for _ in 0..n {
            right = right.next.unwrap();
        }

        while let Some(node) = right.next {
            right = node;
            left = left.next.as_mut().unwrap();
        }

        left.next = left.next.as_mut().unwrap().next.clone();

        dummy.next
    }
}

fn main() {
    let head = Option::from(Box::new(ListNode { val: 1, next: None }));

    let start = Instant::now();
    let result = Solution::remove_nth_from_end(head, 1);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
